aws/table_aws_pricing_product.go

```diff
@@ -0,0 +1,274 @@
+			{Name: "operating_system", Description: "Describes the operating system of your Amazon EC2 instance.", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.operatingSystem")},
+			{Name: "region_code", Description: " A Region is a physical location around the world where data centers are clustered. AWS calls each group of logical data centers an Availability Zone (AZ). Each AWS Region consists of multiple, isolates, and physically separate AZs within a geographical area. The Region code attribute has the same name as an AWS Region, and specifies where the AWS service is available.", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.regionCode")},
+			{Name: "physical_processor", Description: "Describes the processor on your Amazon EC2 instance.", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.physicalProcessor")},
+			{Name: "clock_speed", Description: "Describes the operating speed of your AWS instances.", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.clockSpeed")},
+			{Name: "service_name", Description: "A simplified description about the AWS service.", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.servicename")},
+			{Name: "instance_type", Description: "Describes the instance type, size, and family, which define the CPU, networking, and storage capacity of your instance.", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.instanceType")},
+			{Name: "tenancy", Description: "The type of tenancy allowed on the Amazon EC2 instance.", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.tenancy")},
+			{Name: "license_model", Description: "Describes the license model for your instance.", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.licenseModel")},
+			{Name: "current_generation", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.currentGeneration")},
+			{Name: "pre_installed_software", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.preInstalledSw")},
+			{Name: "processor_architecture", Description: "Describes your processor architecture.", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.processorArchitecture")},
+			// multi az attributes
+			{Name: "deployment_option", Description: "", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.deploymentOption")},
+			// RDS fields
+			{Name: "database_engine", Description: "", Type: proto.ColumnType_STRING, Transform: transform.FromField("Product.Attributes.databaseEngine")},
```

Comment:
Are there other services beside RDS and ElastiCache that have their own fields? If so, do you have a reference list you're using? I'm concerned about scaling this table with specific columns if there are a lot of services/resources with their own specific properties, as this could result in a lot of columns returning `nil` since they'd never have those attributes.

Reply:
You're right ! There are other columns wich are not described in my PR.
I was focused on the most used one's but i think it could be interesting to have all available attributes.

For that, i can use the output of this request : 
``` sql
select
    attribute_name
    ,count(*) as nb
from bedrock_root_us.aws_pricing_service_attribute
group by 1
```

| attribute\_name            | usage |
| -------------------------- | ----- |
| analyticsMode              | 12    |
| awsresource                | 11    |
| baseProductReferenceCode   | 9     |
| bestPractices              | 63    |
| brokerEngine               | 21    |
| bundle                     | 9     |
| cacheType                  | 17    |
| calltype                   | 28    |
| category                   | 10    |
| ciType                     | 11    |
| clientLocation             | 13    |
| computeType                | 4     |
| computeprovider            | 16    |
| deploymentModelDescription | 10    |
| description                | 5     |
| device                     | 10    |
| executionLocation          | 11    |
| extraDayFee                | 27    |
| findingGroup               | 13    |
| flow                       | 10    |
| flowType                   | 18    |
| fromLocation               | 13    |
| fromcountry                | 11    |
| group                      | 645   |
| groupDescription           | 5     |
| integratingApi             | 13    |
| location                   | 28    |
| lookoutvisionimage         | 10    |
| mailboxStorage             | 8     |
| meterMode                  | 8     |
| minimumStorageVolume       | 13    |
| ondemand                   | 11    |
| operation                  | 890   |
| operationType              | 10    |
| opsItems                   | 24    |
| processorFeatures          | 25    |
| productType                | 11    |
| querytype                  | 14    |
| regionCode                 | 5     |
| resourceAssessment         | 10    |
| resourceType               | 95    |
| scanType                   | 12    |
| servicecode                | 8     |
| servicename                | 383   |
| storageClass               | 6     |
| tenancySupport             | 1     |
| termType                   | 6     |
| transactionType            | 1     |
| type                       | 9     |
| usagetype                  | 23    |
| vcpu                       | 2     |
| videoCodec                 | 16    |
| videoResolution            | 8     |
| vmwareregion               | 17    |
| volumeType                 | 1     |
| withActiveUsers            | 9     |

If you're ok with that, I can do the change quickly

Reply:
Ok i was on a good way. It seems the `aws_pricing_service_attribute` has a bug with a pointer.

Some column seems to be duplicated whereas the json returned by aws is valid.

I will take a look on that, and maybe use the exported attributes for my table. It seems to be a lot of them :pleading_face:

Reply:
Fixed in https://github.com/turbot/steampipe-plugin-aws/pull/1372

Reply:
The exhausted list 

| attribute\_name                    | nb  |
| ---------------------------------- | --- |
| 1yearPrice                         | 1   |
| 3yearPrice                         | 1   |
| LeaseContractLength                | 6   |
| OfferingClass                      | 6   |
| PurchaseOption                     | 6   |
| Restriction                        | 48  |
| acceleratorSize                    | 1   |
| acceleratorType                    | 1   |
| accessPointType                    | 1   |
| accessType                         | 1   |
| accountAssistance                  | 3   |
| action                             | 1   |
| activityType                       | 1   |
| acu                                | 1   |
| addonFeature                       | 1   |
| alarmType                          | 1   |
| analyticsMode                      | 1   |
| analyticsType                      | 1   |
| apiCalls                           | 1   |
| apiType                            | 1   |
| apirequestsfree                    | 1   |
| apirequestspaid                    | 1   |
| architecturalReview                | 3   |
| architectureSupport                | 3   |
| attachmentType                     | 2   |
| automaticLabel                     | 1   |
| availability                       | 4   |
| availabilityZone                   | 1   |
| availabilityzone                   | 1   |
| awsiotevents                       | 1   |
| awsresource                        | 1   |
| backupEvaluations                  | 1   |
| backupservice                      | 1   |
| bandwidth                          | 1   |
| baseProductReferenceCode           | 1   |
| bestPractices                      | 3   |
| bitRate                            | 1   |
| brioproductid                      | 1   |
| brokerEngine                       | 1   |
| build                              | 1   |
| bundle                             | 3   |
| bundleDescription                  | 1   |
| bundleGroup                        | 1   |
| cacheEngine                        | 1   |
| cacheMemorySizeGb                  | 1   |
| cacheType                          | 1   |
| cachememorysize                    | 1   |
| callingType                        | 3   |
| calltype                           | 2   |
| capability                         | 1   |
| capacity                           | 1   |
| capacitystatus                     | 2   |
| caseSeverityresponseTimes          | 3   |
| category                           | 2   |
| changerequestsfree                 | 1   |
| changerequestspaid                 | 1   |
| channel                            | 1   |
| chargeid                           | 1   |
| ciType                             | 1   |
| classicnetworkingsupport           | 1   |
| classificationType                 | 1   |
| clientLocation                     | 1   |
| clockSpeed                         | 8   |
| cloudSearchVersion                 | 1   |
| cloudformationresourceProvider     | 1   |
| clustertype                        | 1   |
| comments                           | 1   |
| commitmentDays                     | 1   |
| component                          | 1   |
| compute                            | 1   |
| computeFamily                      | 2   |
| computeType                        | 3   |
| computeprovider                    | 1   |
| computesize                        | 1   |
| computetype                        | 1   |
| concurrencyscalingfreeusage        | 1   |
| connectionType                     | 1   |
| contentSource                      | 1   |
| contentType                        | 2   |
| country                            | 5   |
| countsAgainstQuota                 | 1   |
| cpuArchitecture                    | 1   |
| cputupe                            | 1   |
| cputype                            | 2   |
| currentGeneration                  | 13  |
| customerServiceAndCommunities      | 3   |
| data                               | 1   |
| dataAction                         | 1   |
| dataTransfer                       | 1   |
| dataTransferQuota                  | 1   |
| dataType                           | 1   |
| databaseEdition                    | 1   |
| databaseEngine                     | 3   |
| datastoreStoragetype               | 1   |
| datatransferout                    | 1   |
| datatype                           | 1   |
| dedicatedEbsThroughput             | 5   |
| deploymentLocation                 | 1   |
| deploymentModel                    | 3   |
| deploymentModelDescription         | 1   |
| deploymentOption                   | 4   |
| describes                          | 1   |
| description                        | 26  |
| device                             | 1   |
| deviceDescription                  | 1   |
| deviceName                         | 1   |
| deviceOs                           | 1   |
| deviceProvider                     | 1   |
| deviceType                         | 1   |
| devicename                         | 1   |
| dimension                          | 1   |
| directConnectLocation              | 1   |
| directorySize                      | 1   |
| directoryType                      | 1   |
| directoryTypeDescription           | 1   |
| disableactivationconfirmationemail | 1   |
| dominantnondominant                | 1   |
| durability                         | 3   |
| ebsOptimized                       | 1   |
| ecu                                | 5   |
| edition                            | 1   |
| elasticGraphicsType                | 1   |
| endpoint                           | 1   |
| endpointType                       | 2   |
| engine                             | 2   |
| engineCode                         | 3   |
| engineMediaType                    | 1   |
| enhancedNetworkingSupport          | 1   |
| enhancedNetworkingSupported        | 6   |
| entityType                         | 1   |
| eventType                          | 4   |
| executionFrequency                 | 1   |
| executionLocation                  | 1   |
| executionMode                      | 1   |
| externalInstanceType               | 1   |
| extraDayFee                        | 1   |
| feature                            | 1   |
| feeCode                            | 6   |
| feeDescription                     | 6   |
| feecode                            | 1   |
| fileCacheType                      | 1   |
| fileSystemType                     | 1   |
| findingGroup                       | 1   |
| findingSource                      | 1   |
| findingStorage                     | 1   |
| flow                               | 1   |
| flowType                           | 1   |
| frameRate                          | 1   |
| frameworkType                      | 1   |
| freeDays                           | 1   |
| freeOverage                        | 1   |
| freeQueryTypes                     | 1   |
| freeTier                           | 1   |
| freeTrial                          | 1   |
| freeUsageIncluded                  | 2   |
| freetier                           | 1   |
| frequency                          | 1   |
| frequencyMode                      | 1   |
| fromLocation                       | 12  |
| fromLocationType                   | 12  |
| fromRegionCode                     | 10  |
| fromcountry                        | 3   |
| gb                                 | 1   |
| georegioncode                      | 1   |
| gets                               | 1   |
| gpu                                | 3   |
| gpuMemory                          | 2   |
| granularity                        | 1   |
| graphqloperation                   | 1   |
| groundstation                      | 1   |
| group                              | 66  |
| groupDescription                   | 52  |
| groupDetails                       | 1   |
| highAvailability                   | 1   |
| includedEvents                     | 1   |
| includedServices                   | 3   |
| indexingSource                     | 1   |
| ingestType                         | 1   |
| input                              | 1   |
| inputMode                          | 2   |
| inputType                          | 1   |
| inputtype                          | 1   |
| insightstype                       | 1   |
| instance                           | 1   |
| instanceCapacity10xlarge           | 1   |
| instanceCapacity12xlarge           | 1   |
| instanceCapacity16xlarge           | 2   |
| instanceCapacity18xlarge           | 1   |
| instanceCapacity24xlarge           | 1   |
| instanceCapacity2xlarge            | 2   |
| instanceCapacity32xlarge           | 1   |
| instanceCapacity4xlarge            | 2   |
| instanceCapacity8xlarge            | 2   |
| instanceCapacity9xlarge            | 1   |
| instanceCapacityLarge              | 1   |
| instanceCapacityMedium             | 1   |
| instanceCapacityMetal              | 1   |
| instanceCapacityXlarge             | 2   |
| instanceFamily                     | 13  |
| instanceFunction                   | 1   |
| instanceName                       | 1   |
| instanceStorageGb                  | 1   |
| instanceType                       | 21  |
| instanceTypeFamily                 | 2   |
| instances                          | 1   |
| instancesku                        | 1   |
| instancetype                       | 1   |
| integratingApi                     | 2   |
| integratingFeature                 | 2   |
| integratingService                 | 2   |
| intelAvx2Available                 | 2   |
| intelAvxAvailable                  | 2   |
| intelTurboAvailable                | 2   |
| io                                 | 1   |
| ioRequestType                      | 1   |
| isSpotInstance                     | 1   |
| iscommitcpsku                      | 1   |
| isfree                             | 1   |
| isshadow                           | 1   |
| iswebsocket                        | 1   |
| jobFee                             | 1   |
| jobnshipp                          | 1   |
| labelingTaskType                   | 1   |
| launchSupport                      | 3   |
| learningMode                       | 1   |
| license                            | 1   |
| licenseModel                       | 6   |
| licenseType                        | 3   |
| lineType                           | 1   |
| location                           | 197 |
| locationType                       | 197 |
| lockeprofiles                      | 1   |
| logsDestination                    | 1   |
| logsSource                         | 1   |
| logsType                           | 1   |
| lookoutvisionimage                 | 1   |
| machineLearningProcess             | 1   |
| mailboxStorage                     | 1   |
| marketoption                       | 1   |
| maxIopsBurstPerformance            | 1   |
| maxIopsvolume                      | 1   |
| maxThroughputvolume                | 2   |
| maxVolumeSize                      | 5   |
| maximumCapacity                    | 2   |
| maximumExtendedStorage             | 1   |
| maximumStorageVolume               | 1   |
| memory                             | 17  |
| memoryGib                          | 4   |
| memorygb                           | 1   |
| memorytype                         | 2   |
| messageDeliveryFrequency           | 1   |
| messageDeliveryOrder               | 1   |
| messageType                        | 1   |
| messagetype                        | 2   |
| meterMode                          | 1   |
| meteringType                       | 1   |
| meterunit                          | 1   |
| minVolumeSize                      | 4   |
| minimumStorageVolume               | 1   |
| networkPerformance                 | 11  |
| newcode                            | 2   |
| nlpclass                           | 1   |
| normalizationSizeFactor            | 6   |
| numbertype                         | 3   |
| offer                              | 1   |
| ondemand                           | 1   |
| operatingSystem                    | 9   |
| operation                          | 105 |
| operationOutcome                   | 1   |
| operationType                      | 1   |
| operationsSupport                  | 3   |
| opsItems                           | 1   |
| origin                             | 1   |
| os                                 | 1   |
| osLicenseModel                     | 1   |
| outpostusage                       | 1   |
| output                             | 1   |
| outputMode                         | 2   |
| outputType                         | 1   |
| overageType                        | 1   |
| overhead                           | 1   |
| parameterType                      | 1   |
| physicalCores                      | 1   |
| physicalCpu                        | 1   |
| physicalGpu                        | 1   |
| physicalProcessor                  | 6   |
| pipeline                           | 2   |
| planType                           | 1   |
| platofeaturetype                   | 1   |
| platoinstancename                  | 1   |
| platopagedatatype                  | 1   |
| platoprotectionpolicytype          | 1   |
| platousagetype                     | 1   |
| portSpeed                          | 1   |
| power                              | 1   |
| preInstalledSw                     | 2   |
| predictionType                     | 1   |
| pricingUnit                        | 9   |
| pricingplan                        | 1   |
| primaryplaceofuse                  | 4   |
| proactiveGuidance                  | 3   |
| processorArchitecture              | 7   |
| processorFeatures                  | 6   |
| productFamily                      | 193 |
| productName                        | 1   |
| productSchemaDescription           | 1   |
| productType                        | 2   |
| productgroupid                     | 1   |
| productsubgroup                    | 1   |
| producttype                        | 1   |
| programmaticCaseManagement         | 3   |
| protocol                           | 2   |
| provider                           | 2   |
| provisioned                        | 2   |
| querytype                          | 1   |
| queueType                          | 1   |
| readtype                           | 1   |
| realtimeoperation                  | 1   |
| recipient                          | 1   |
| regionCode                         | 158 |
| releaseType                        | 1   |
| replicationType                    | 2   |
| requestDescription                 | 1   |
| requestType                        | 3   |
| reserveType                        | 1   |
| reservedTier                       | 1   |
| resolution                         | 1   |
| resource                           | 1   |
| resourceAssessment                 | 1   |
| resourceEndpoint                   | 2   |
| resourcePriceGroup                 | 1   |
| resourceType                       | 3   |
| response                           | 1   |
| rootvolume                         | 1   |
| routingTarget                      | 2   |
| routingType                        | 2   |
| runningMode                        | 1   |
| scanType                           | 1   |
| serverLocation                     | 1   |
| servicecode                        | 199 |
| servicename                        | 186 |
| simulatorName                      | 1   |
| sitelinkOriginPopGroup             | 1   |
| size                               | 1   |
| snapshotarchivefeetype             | 1   |
| snowballType                       | 2   |
| softwareIncluded                   | 1   |
| softwareType                       | 1   |
| sourcetype                         | 1   |
| ssd                                | 1   |
| standardGroup                      | 1   |
| standardStorage                    | 1   |
| standardStorageRetentionIncluded   | 1   |
| steps                              | 1   |
| storage                            | 13  |
| storageCapacityQuota               | 1   |
| storageClass                       | 8   |
| storageDescription                 | 2   |
| storageFamily                      | 1   |
| storageMedia                       | 10  |
| storageType                        | 4   |
| storageclass                       | 2   |
| storagetype                        | 4   |
| subcategory                        | 1   |
| subscriptionType                   | 1   |
| subservice                         | 1   |
| supportedModes                     | 2   |
| technicalSupport                   | 3   |
| tenancy                            | 4   |
| tenancySupport                     | 1   |
| termType                           | 199 |
| thirdpartySoftwareSupport          | 3   |
| throughput                         | 1   |
| throughputCapacity                 | 1   |
| throughputClass                    | 1   |
| tickettype                         | 1   |
| tier                               | 4   |
| tiertype                           | 1   |
| toLocation                         | 12  |
| toLocationType                     | 13  |
| toRegionCode                       | 9   |
| tocountry                          | 5   |
| trafficDirection                   | 1   |
| training                           | 3   |
| transactionType                    | 3   |
| transactiontype                    | 1   |
| transcodingResult                  | 2   |
| transferType                       | 12  |
| trialProduct                       | 1   |
| type                               | 8   |
| updates                            | 1   |
| upfrontCommitment                  | 1   |
| usageFamily                        | 1   |
| usageTier                          | 1   |
| usageVolume                        | 1   |
| usagetype                          | 199 |
| usertype                           | 1   |
| uservolume                         | 1   |
| vCpu                               | 1   |
| vcpu                               | 20  |
| version                            | 5   |
| videoCodec                         | 1   |
| videoFrameRate                     | 1   |
| videoMemoryGib                     | 1   |
| videoQuality                       | 1   |
| videoQualitySetting                | 1   |
| videoResolution                    | 2   |
| virtualInterfaceType               | 1   |
| vmwareproductid                    | 1   |
| vmwareregion                       | 1   |
| volumeApiName                      | 1   |
| volumeName                         | 1   |
| volumeType                         | 16  |
| vpcnetworkingsupport               | 1   |
| whoCanOpenCases                    | 3   |
| withActiveUsers                    | 1   |
| workforceType                      | 2   |

Reply:
@aminvielledebatAtBedrock Do you also know what the breakout of these attributes are per service? My concern is that if there are a large number of services, and some attributes are only present in a few services each, that if a user runs `select * from aws_pricing_product where service_code = '...'`, they may get back a lot of columns with `nil` data, which is a lot to sort through, and may cause confusion, i.e., is this attribute expected to have data and the table didn't return it correctly?

For instance, `volume_name` will only be in the EC2/EBS product most likely, but will be `nil` for all others. With the updated listed, it seems like there are a lot of attributes that only appear once, so there will be a lot of `nil` column values coming through consistently.

Not sure if this is the right direction, but I wonder what if we broke up the `aws_pricing_product` table into multiple tables, one per service code? It'd be a lot of tables, but would that help with reducing the number of `nil` attributes?

Alternatively, we can avoid adding any specific attributes as columns and rely on users to pull them out, e.g., `select attributes -> 'volume_name' as volume_name from aws_pricing_product where service_code = '...'`, while leaving the common attributes, like `usage_type` in as columns.

Reply:
I don't know what's the better option, if we take account the following points : 
- Only 9 attributes (on 417) are used by the most part of aws services (for 199 referenced services)
  There are too common, and don't allow the end user to filter the results
  For example, AmazonElasticCache on `eu-west-3` returns 861 rows :disappointed: 
  The only difference between rows will be the json attributes columns. Not so good IMO.
- Attributes are used by the API for filtering. Not using them involves fetching all data per `service_code` 
- 417 attributes is too much for one table. I agree with you.
- We are talking about 199 different services. I really like steampipe, but not enough to write 199 tables :laughing: 

Do you think dynamic tables could be possible (with a first call to `DescribeServices` to fetch column definitions) ? 

Or the least bad could be keeping only 9 attributes, with an json attributes column... 

I'll follow you !

Reply:
@aminvielledebatAtBedrock Thanks for the figures, they're really helpful in understanding what the API returns and how the data looks across the services. Let me have a look through the data and service codes again, and I'll respond here again tomorrow with any findings/suggestions I have.

Overall, my thoughts are:
- One table seems like the best design for now. We could create 199+ tables dynamically, but I'm not sure I'd want to flood the schema with all of those tables.
- With 417 attributes, that's a lot to try map manually into columns. I'm not sure if there's a way to map automatically without making the API calls (maybe there's a reference in the AWS SDK code though, like there are endpoints).
- We could still implement filtering by attribute by adding a `filters` JSON column that serves as a `ListConfig` key column. For instance, taking one example from your current table, it would then look something like:
```
select
  service_code,
  attributes -> 'instanceType' as instance_type,
  attributes -> 'instanceName' as instance_name,
  price_per_unit::numeric::money
from
  aws_pricing_product 
where
  service_code = 'AmazonEC2' 
  and region_code = 'eu-west-3' 
  and location_type = 'AWS Region' 
  and filters = '{
    "instanceType": "c5.2xlarge",
    "operatingSystem": "Linux",
    "tenancy": "Shared",
    "preInstalledSw": "NA"
    "capacityStatus": "Used"
  }'::jsonb;
  ```
  - The `service_code`, `region_code`, and `location_type` columns would be actual columns (based on the counts above), while the others would be passed through the new `filters` column. If key columns are passed in along with `filter`, like in the example above, they should all be combined into a single `input.Filters`.
  - Note: I actually prefer using `jsonb_build_object` instead, but there's currently a bug in the Steampipe SDK preventing that from working correctly with JSON type key columns.
  
  I believe the design above would give users the ability to query pricing products, top column support for common columns, and the flexibility to pass in other attributes for filtering.
  
  @aminvielledebatAtBedrock @johnsmyth - Curious to hear your thoughts and feedback, thanks!

Reply:
@cbruno10 I think that your approach seems reasonable, but it will be pretty hard for people to use.  I like @aminvielledebatAtBedrock 's idea of dynamic tables, but we would end up with alot of dynamic tables.   So maybe we can do both:
- add the 'generic' table to the plugin as @cbruno10 suggested
- add the per-service tables (dynamic tables if that works, else manually curated) but behind a config argument:
    - These tables would NOT be enabled by default
    - The argument would ideally accept an array of table (service) names, and allow the '*' wildcard.  (similar to the how the `regions` arg works)
        - This would allows users to only enable the tables they want
        - This would also allow user to enable it in only a single connection, which is probably desirable when using an aggregator (otherwise you would get the same pricing info from every connection in the aggregator).
        
thoughts?

Reply:
@johnsmyth I like the idea of putting the dynamic tables behind a config argument, and agree with the other design notes in your comment.

For the scope of this PR, I think we could start with the generic table, and then create an issue to track adding dynamic tables, as that would be considerable more work to add, e.g., decide on table and column naming structure that works for all service/service codes, test API limits in case someone had `*`, meaning they want all tables (each of which would require a separate API call I believe, as you can only pass 1 service code per request), test the time it takes to create the tables.

Reply:
Done. Could you have a look @cbruno10 please ?